perf(compiler): reduce amount of generated code for safe accesses and nullish coalescing

crisbeto opened the pull request:
This is follow-up from #41437 and it reduces the amount of code we generate for safe property accesses (`a?.b`) and nullish coalescing (`a ?? b`) by:
1. Reusing variables in nested nullish coalescing expressions.
2. Not initializing temporary variables to `null`. The way our code is generated means that the value will always be overwritten before we compare against it so the initializer didn't really matter.

Fixes #41491.

petebacondarwin commented on packages/compiler-cli/test/compliance/test_cases/r3_view_compiler/nullish_coalescing/nullish_coalescing_host_bindings.js:
```diff
@@ -1,13 +1,12 @@
-      let tmp_b_0 = null;
-      let tmp_b_1 = null;
-      return ctx.logLastName((tmp_b_0 = (tmp_b_1 = ctx.lastName) !== null && tmp_b_1 !== undefined ? tmp_b_1 : ctx.lastNameFallback) !== null && tmp_b_0 !== undefined ? tmp_b_0 : "unknown");
+      let tmp_b_0;
```
Note that if you wrap variable names (in these test expectation files) in `$` symbols then they are resilient to change. The checker will just ensure that the same identifier appears in each instance but not that it matches exactly. E.g. `let $tmp_b_0$;`, etc.

petebacondarwin commented on packages/compiler-cli/test/compliance/test_cases/r3_view_compiler/nullish_coalescing/nullish_coalescing_interpolation_template.js:
```diff
@@ -8,12 +8,11 @@ template: function MyApp_Template(rf, ctx) {
     i0.ɵɵtextInterpolate1("Hello, ", (tmp_0_0 = ctx.firstName) !== null && tmp_0_0 !== undefined ? tmp_0_0 : "Frodo", "!");
     i0.ɵɵadvance(2);
-    i0.ɵɵtextInterpolate1("Your last name is ", (tmp_1_0 = (tmp_1_1 = ctx.lastName) !== null && tmp_1_1 !== undefined ? tmp_1_1 : ctx.lastNameFallback) !== null && tmp_1_0 !== undefined ? tmp_1_0 : "unknown", "");
+    i0.ɵɵtextInterpolate1("Your last name is ", (tmp_1_0 = (tmp_1_0 = ctx.lastName) !== null && tmp_1_0 !== undefined ? tmp_1_0 : ctx.lastNameFallback) !== null && tmp_1_0 !== undefined ? tmp_1_0 : "unknown", "");
```
Could we even go further and reuse `tmp_0_0` too?

crisbeto replied:
We could, at least for the nullish coalescing, but the temporary variable generation logic is used for other kinds of expressions and I don't know what the repercussions would be to change it. I think that these changes get us most of the way there in terms of reducing the variable size, because they allow minifiers to collapse the declaration to a single line. E.g.

```
let a;
let b;
```

Can be minified to `let a, b`.

crisbeto committed:
perf(compiler): reduce amount of generated code for safe accesses and nullish coalescing

This is follow-up from #41437 and it reduces the amount of code we generate for safe property accesses (`a?.b`) and nullish coalescing (`a ?? b`) by:
1. Reusing variables in nested nullish coalescing expressions.
2. Not initializing temporary variables to `null`. The way our code is generated means that the value will always be overwritten before we compare against it so the initializer didn't really matter.

Fixes #41491.

AndrewKushnir commented:
[Presubmit](http://test/OCL:368116513:BASE:368116528:1618273127633:d5a92714).

AndrewKushnir reviewed (approved):
LGTM, thanks for the improvement @crisbeto 👍 

Quick question: could you please describe why the `risk: medium` label is applied, should caretakers handle the merging of this PR in a spacial way (like merge and sync into g3 separately from other changes)?

crisbeto commented:
I marked it as `risk: medium`, because removing the initializer from temporary variables affects other kinds of syntax like safe property reads. I was thinking that we may want to run a global presubmit just in case.

AndrewKushnir commented:
Thanks @crisbeto, I will run global TAP tonight as well and let you know.

AndrewKushnir commented:
[Global presubmit](https://test.corp.google.com/OCL:368116513:BASE:368340655:1618365726401:14efa264).

AndrewKushnir commented:
@crisbeto FYI, presubmits (including the global one) are successful for the changes in this PR.
